### ISSUE TITLE
What is the possible op substitute for set of OPerations on CPU to DSP Hexagon

### ISSUE THREAD

**Issue by kzos** [tensorflow/tensorflow#10763]
OS: Ubuntu 16.04 64bits
Android Version: 7.1 (Nougat)
NDK Version: android-ndk-r12b
HEXAGON SDK: 3.1
nnlib source: https://source.codeaurora.org/quic/hexagon_nn/nnlib

I am trying to build a graph to run on hexagon,
it uses few op's which hexagon doesn't support.

native : graph_transferer.cc:109 Failed to transfer graph Invalid argument: Mean has not been implemented yet.

OP's details are as below:-

 {"Mean", SupportedOpType::**??**},
 {"RealDiv", SupportedOpType::**??**},
 {"Pow", SupportedOpType::**??**},
 {"Conv2DBackpropInput", SupportedOpType::**??**},
 {"Square", SupportedOpType::**??**},
 {"SquaredDifference", SupportedOpType::**??**},
 {"StopGradient", SupportedOpType::**??**},
 {"Reciprocal", SupportedOpType::**??**},

Is there any possible substitute for the operation types Mean, Pow, Conv2DBackpropInput etc.. for hexagon.

thanks in advance.

**Comment by petewarden** [comment]
Unfortunately the current approach to HVX acceleration requires new op implementations to be written for the DSP before they can be used. Looking at the list of ops, a lot of them seem like they might be related to batch normalization. Are you able to run the `fold_constants` transform using the graph transform tool and see if some of them disappear?

**Comment by kzos** [comment]
With **fold_constants, quantize option,** for tranform_graph,
On Hexagon, was able to get through <registration of nodes>.
Failed in <Setupgraph> during instantiating the graph.

```
hexagon/src/newnode.c:177:### append_const_node
 hexagon/src/graphops.c:60:CAUSALITY VIOLATION: node ea63a670 (id=0x1005c) referenced output of node ea63d210 (id=0x10052) before instantiated in the graph
 hexagon/src/graphops.c:60:CAUSALITY VIOLATION: node ea63a910 (id=0x1005f) referenced output of node ea63d210 (id=0x10052) before instantiated in the graph
 hexagon/ops/src/op_prefree.c:71:prefree node id 0 ctor
 hexagon/src/newnode.c:177:### append_const_node
```

Logging in nnlib is not friendly, no node names etc..
I will try to debug it further to get names stuff and try to fix (**Comment if you got any pointers to debug here**).

As I couldn't get working on Hexagon, to understand the changes underwent in graph after using **fold_constants,** I ran the graph on CPU,
all ops are getting substituted with moments, as below:(**What it does?**)

```
native : stat_summarizer.cc:407 	              Dequantize	   51.109	    0.723	    0.705	  0.506%	 36.654%	  2457.600	Conv2D_2
native : stat_summarizer.cc:407 	                    Mean	   51.828	    0.848	    1.080	  0.775%	 37.429%	     0.128	moments_1/Mean
native : stat_summarizer.cc:407 	            StopGradient	   52.920	    0.014	    0.015	  0.010%	 37.440%	     0.000	moments_1/StopGradient
```


I assume, fold_constants can be a safe bet to get graph working on hexagon once after fixing the instantiate graph part.(Comment? if you have anything to say)

**Can I visualize pre trained graph(pb) by any means if not through tensorboard?**
that would help me debug.

thanks,

**Comment by kzos** [comment]
@petewarden,
I am implementing few of the ops mentioned above.
started with mean, stop gradient and then
conv2dbackpropinput.


For op_mean, collecting information from graph transferrer as below(4 inputs) (1 output):

      const struct tensor *input = self->input[0];
      const struct tensor *axes = self->input[1];
      const struct tensor *dims = self->input[2];
      const struct tensor *reduction = self->input[3];
      struct tensor *output = self->outputs[0];
      //do mean operation based on axes, dims

      
I am not sure what does **reduction_indices** does. -> I assume it is reducing the order of matrix from 3 to (3-2)->1

![screenshot from 2017-07-06 19-58-10](https://user-images.githubusercontent.com/5010954/27916202-fdb39268-6285-11e7-8c79-d15ec5d46a63.png)

number of inputs vary after graph transformation.
so is there a guide to know number of inputs as well outputs for operations like -> Mean, Stop gradient, etc.

Thanks,

**Comment by petewarden** [comment]
A good place to look for op definitions is here:

https://github.com/tensorflow/tensorflow/tree/master/tensorflow/core/ops

It's where all the different ops have their inputs, outputs, and attribute signatures defined. For example, here's the signature for "Mean":
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/math_ops.cc#L1154
```
REGISTER_OP("Mean")
  .Input("input: T")
  .Input("reduction_indices: Tidx")
  .Output("output: T")
  .Attr("keep_dims: bool = false")
  .Attr("T: numbertype")
  .Attr("Tidx: {int32, int64} = DT_INT32")
```
This is saying that the Mean op takes two inputs, the first of which is any number type, and the second must be an int32 or int64. It also has an option to keep the dimensions. The doc string below the definition has more information about what it does.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.
